radsat-sk/operation/subsystems/RDosimeter.c

```diff
@@ -1,111 +1,235 @@
- *	data from the experimental Melanin-Dosimeter boards. Relies on asynchronous
- *	I2C calls to send commands and receive the information in a response.
+ * Iterates through each channel on both boards to collect and store payload
+ * data from the experimental Melanin-Dosimeter boards. Relies on I2C calls to
+ * send commands and receive the information in a response.
  *
- *	@header	"software-and-command/payload_processing/payload_processing.h"
- *	@param	void
- *	@pre	Melanin-Dosimeter boards are powered on. No initialization necessary
- * 	@post	Payload data is stored in non-volatile memory (TODO: location TBD)
- * 	@return	void
+ * @pre I2C must be initialized
+ * @post If successful, sends new formatted group of readings to the Downlink Manager
+ * @param void
+ * @return Returns 0 on Success, anything else indicates error (e.g. from I2C).
+ * 		   In the event of error, no data will have been sent to the Downlink
+ * 		   Manager.
  */
-void requestReadingsAllChannels( void )
+int dosimeterCollectData(void)
 {
+	// internal buffer for receiving I2C responses
+	uint8_t dataResponse[DOSIMETER_RESPONSE_LENGTH] = { 0 };
+
+	// TODO: prepare a protobuf struct to populate within the nested for loops
+
 	// iterate through both melanin-dosimeter boards
-	for ( uint8_t dosimeterBoard = 0; dosimeterBoard < DOSIMETER_COUNT; dosimeterBoard++ ) {
+	for (uint8_t dosimeterBoard = dosimeterBoardOne; dosimeterBoard < dosimeterBoardCount; dosimeterBoard++) {
+
+		// request data from each sensor on a particular board
+		for (uint8_t adcChannel = adcChannelOne; adcChannel < adcChannelCount; adcChannel++) {
 
-    	// request data from each sensor on a particular board
-		for ( uint8_t adcChannel = 0; adcChannel < 8; adcChannel++ ) {
+			// reset internal buffer
+			memset(dataResponse, 0, DOSIMETER_RESPONSE_LENGTH);
 
-			// write command to dosimeter board using i2c
-			uint8_t i2cResult = i2c->writeByte( dosimeterBoardSlaveAddr[dosimeterBoard],
-												dosimeterCommandBytes[adcChannel]
-											  	);
+			// tell dosimeter to begin conversion; receive 12-bit data into our internal buffer
+			int error = i2cTalk(dosimeterBoardSlaveAddr[dosimeterBoard], DOSIMETER_COMMAND_LENGTH,
+								DOSIMETER_RESPONSE_LENGTH, &dosimeterCommandBytes[adcChannel],
+								dataResponse, DOSIMETER_I2C_DELAY);
 
 			// check for success of I2C command
-			if ( i2cResult != 0 ) {
-				// 0 = success code; anything else is a failure
-				// TODO: replace assertion with logging and proper error handling
-				ASSERT("ERROR");
+			if ( error != 0 )
+				return error;
+
+			float finalVoltage = convertCountsToVoltage(dataResponse[0], dataResponse[1]);
+
+			// if reading the temperature sensor, convert it to celsius
+			if (adcChannel == temperatureSensor) {
+				uint16_t temperature = convertVoltageToTemperature(finalVoltage);
 			}
 
-			// TODO: Do we need to add any delays? Don't think so, but should confirm
+			// TODO: format data into Protobuf message

```

Comment:
Would just like to confirm that this protobuf message will essentially have sections and each section will be a reading from each channel. 

(Excuse my crude diagram)

----Dosimeter Board Protobuf Message---
           |                              |                      ...
Channel 0 data      Channel 1 data          ...

Again, my understanding of protobuf is close to none but I would like to confirm that all the data collected will be sent down as one message.

Then we downlink the protobuf message with all the channels, including the temperature data. I would also like to confirm that we have the dosimeterTemperature function because we will implement this function in safe mode, or was there another reason for the dosimeterTemperature function.

Reply:
Yeah, the downlinked protobuf message will be sent as one struct that contains all of the data, organized into nice little variables